File: examples/detectingStampPattern.cpp
#include "Vernier.hpp"

using namespace vernier;
using namespace cv;
using namespace std;

/** This example shows how to estimate the pose of stamp markers in an image.
 */
int main() {

    // Loading the image
    string filename = "stampPatternImage_15um.jpg";
    Mat image = cv::imread(filename);

    // Detecting the pattern and estimating its pose
    double physicalPeriod = 15; // µm
    int snapshotSize = 820; // pixels
    StampPatternDetector detector(physicalPeriod, "stampPattern.png", snapshotSize);
    detector.compute(image);

    // Displaying the pose if a pattern has been found
    cout << "Detector: " << detector.toString() << endl;
    if (detector.patternFound()) {
        cout << "Found " << detector.markers.size() << " markers in " << filename << endl;
        for (auto it : detector.markers) {
            cout << "  Stamp " << it.first << " at " << it.second.toString() << endl;
        }
        detector.draw(image);
    } else {
        cout << "Marker not found..." << endl;
    }

    // Showing image and its spectrum
    detector.showControlImages();
    imshow(filename, image);
    waitKey(3000);
}

File: python/examples/detectingStampPattern.py
#
# This file is part of the VERNIER Library.
#
# Copyright (c) 2018-2025 CNRS, ENSMM, UMLP.
#

"""Python port of examples/detectingStampPattern.cpp.

This example shows how to estimate the pose of stamp markers in an image.

Run it from the build directory, where the module and the images live:

    cd build/python && python3 detectingStampPattern.py [--show]
"""

import sys

import pyvernier as vernier

# Loading the image
filename = "stampPatternImage_15um.jpg"
image = vernier.readImage(filename)

# Detecting the pattern and estimating its pose
physicalPeriod = 15  # µm
snapshotSize = 820  # pixels
detector = vernier.StampPatternDetector(physicalPeriod, "stampPattern.png", snapshotSize)
detector.compute(image)

# Displaying the pose if a pattern has been found
print("Detector:", detector)
if detector.patternFound():
    # `markers` is a dict {stamp id: Pose}.
    print("Found %d markers in %s" % (len(detector.markers), filename))
    for stamp_id, pose in sorted(detector.markers.items()):
        print("  Stamp %d at %s" % (stamp_id, pose))
    image = detector.draw(image)
else:
    print("Marker not found...")

# Showing image and its spectrum (needs a display, hence the flag)
if "--show" in sys.argv:
    detector.showControlImages()
    vernier.showImage(filename, image)
    vernier.waitKey(3000)
